### DISCUSSION TITLE
Deployment Issue on bridge

### DISCUSSION THREAD

**Discussion by akyo8** [zkSync-Community-Hub/zksync-developers#82] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

0.4.0

### zksync-web3 Version

0.14.3

### Hardhat.config.ts

import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";

import "@matterlabs/hardhat-zksync-verify";

import dotenv from "dotenv";
dotenv.config();

// dynamically changes endpoints for local tests
const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://localhost:3050",
        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://zksync2-testnet.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
        // contract verification endpoint
        verifyURL:
          "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {},
  },
  defaultNetwork: "zkTestnet",
  // networks: {
  //   hardhat: {
  //     zksync: false,
  //   },
  //   zkSyncTestnet,
  // },
  networks: {
    goerli: {
      url: "https://goerli.infura.io/v3/056f9ac6b8e2455a9492f68a7053e7b0", // The Ethereum Web3 RPC URL (optional).
    },
    zkTestnet: {
      url: "https://testnet.era.zksync.dev", // The testnet RPC URL of zkSync Era network.
      ethNetwork: "goerli", // The Ethereum Web3 RPC URL, or the identifier of the network (e.g. `mainnet` or `goerli`)
      zksync: true,
    },
  },
  solidity: {
    version: "0.8.17",
  },
};

export default config;

### Deployment Script (WITHOUT PRIVATE KEY)

import { Wallet, Provider } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";

// load env file
import dotenv from "dotenv";
dotenv.config();

// HTTPS RPC endpoint
const MAINNET_RPC_ENDPOINT =
  "https://goerli.infura.io/v3/056f9ac6b8e2455a9492f68a7053e7b0";

// Token address
const TOKEN_ADDRESS = "0x4341Ff79E0d988D76BfE83CA081624d77f08764E";

// Amount of tokens
const AMOUNT = "5000";

const WALLET_PRIV_KEY = process.env.WALLET_PRIV_KEY || "";

if (!WALLET_PRIV_KEY) {
  throw new Error("Wallet private key not configured in env file");
}

export default async function (hre: HardhatRuntimeEnvironment) {
  console.log(`Running script to bridge ERC20 to L2`);

  // Initialize the wallet.
  const provider = new Provider(
    // @ts-ignore
    MAINNET_RPC_ENDPOINT || hre.config.networks.zkSyncTestnet.ethNetwork
  );

  const wallet = new Wallet(WALLET_PRIV_KEY, provider);

  // Create deployer object
  const deployer = new Deployer(hre, wallet);

  // Deposit ERC20 tokens to L2
  const depositHandle = await deployer.zkWallet.deposit({
    to: deployer.zkWallet.address,
    token: TOKEN_ADDRESS,
    amount: ethers.utils.parseEther(AMOUNT), // assumes ERC20 has 18 decimals
    // performs the ERC20 approve action
    approveERC20: true,
  });

  console.log(`Deposit transaction sent ${depositHandle.hash}`);
  console.log(`Waiting for deposit to be processed in L2...`);
  // Wait until the deposit is processed on zkSync
  await depositHandle.wait();
  console.log(`ERC20 tokens available in L2`);
}

### Contract Code

https://goerli.etherscan.io/address/0x4341Ff79E0d988D76BfE83CA081624d77f08764E

### Does this work on other EVMs? (If yes, please list at least 1 of them)

https://goerli.etherscan.io/address/0x4341Ff79E0d988D76BfE83CA081624d77f08764E

### Description of What Your Contract Does

Token Contract

**Comment by dutterbutter** [comment]
@akyo8 what is the error here? Are you sure that the `zkWallet` has an ERC20 token balance enough to cover the transfer amount?

**Comment by akyo8** [comment]
@dutterbutter It would be great if you can give me a brief instruction to let me know how can I update ERC20 token balance in zkWallet. Do I need any specific deployment for it?

**Reply by dutterbutter** [reply, marked as answer]
Hey @akyo8 please refer to the documentation that outlines how to interact with [accounts](https://era.zksync.io/docs/api/js/accounts.html) and make [L1-L2 transactions](https://era.zksync.io/docs/api/js/accounts-l1-l2.html).

You need to ensure your wallet has appropriate funds before being able to deploy, send, bridge etc.

**Comment by akyo8** [comment]
@dutterbutter thanks for your update. here is my quick question. 
Is subgraph support zkSync era testnet? I can see zkSync era but not able to see testnet mark. So I want to make sure if zkSync era testnet supported for subgraph. And if we want to implmeent subgraph for testnet what could be ideal solution

**Reply by dutterbutter** [reply]
TheGraph subgraphs are not yet supported on zkSync Era testnet but we are working on inclusion.

**Reply by uF4No** [reply]
actually subgraphs are supported, just that the graph-cli does not fully support zkSync yet (we're working on it!) so you have to create them manually for now. We'll have the fix for the graph cli very soon.

**Comment by akyo8** [comment]
Thanks for your reply. So when do you think if it support zkSync era testnet. It is very important for us for launch schedule

**Reply by uF4No** [reply]
In the next release after [this PR](https://github.com/graphprotocol/graph-tooling/pull/1418) is merged

**Reply by uF4No** [reply]
@akyo8 [theGraph cli version 0.55.0](https://www.npmjs.com/package/@graphprotocol/graph-cli?activeTab=readme) now supports both zkSync Era testnet and mainnet

**Comment by vatosherrod** [comment]
The provided configuration and deployment script appear to be set up for deploying ERC20 tokens to the zkSync Layer 2 network using the zksync-web3 library and hardhat-zksync-deploy plugin. The deployment script initializes a wallet, approves ERC20 tokens for deposit, and deposits the specified amount of tokens to the zkSync network. The contract code is not provided, but the token contract address on the Goerli testnet is referenced. Overall, the setup seems appropriate for bridging ERC20 tokens to Layer 2 on zkSync.

**Comment by beckyhaynes** [comment]
Your configuration and deployment script seem to be set up correctly for deploying an ERC20 token contract to zkSync's testnet. However, I noticed a couple of things:

Environment Variables: You're using environment variables for the wallet private key and the RPC endpoint. Ensure that these variables are correctly set in your environment or in an .env file.

Approval: You're approving the ERC20 token transfer within the deposit function. This is correct, but ensure that the wallet has sufficient allowance for the deposit amount before calling this script.

Verification: After deploying the contract, you might want to verify it on zkSync's block explorer for transparency and validation purposes. You can use the provided verification endpoint in your Hardhat config to automate this process.

Overall, your setup looks good for deploying and interacting with ERC20 tokens on zkSync's testnet. Make sure to handle environment variables securely and verify the deployed contract for added confidence.

**Comment by phicurby** [comment]
To bridge ERC20 tokens to L2 on zkSync, you can use the provided deployment script and configuration in your Hardhat environment. Ensure that you have set up your environment variables correctly, including the wallet private key. The script deposits ERC20 tokens to L2 and waits for the deposit transaction to be processed. After the deposit is completed, the tokens should be available in L2. You can verify the deployed contract on the Goerli testnet using Etherscan.

**Comment by jwitte3** [comment]
It seems there might be an issue with your deployment script or configuration. Please ensure that the network settings and provider URLs are correctly configured. Additionally, make sure that your wallet has sufficient funds for the transaction and that the private key is properly set in the environment variables.
